### PR TITLE
Fix lib: interface for GroundedSchema

### PR DESCRIPTION
**Pull request by Necr0x0Der**
- setLocalFunc is renamed to setLocalSchema (to add setLocalPredicate in future)
- freeing memory from the pointer returned by "lib:" GroundedSchema is added
- unit test for setLocalSchema is added

### CODE REVIEW

File: opencog/atoms/execution/ExecutionOutputLink.cc [hunk 1]
```diff
@@ -261,7 +266,7 @@ void LibraryManager::setLocalFunc(std::string libName, std::string funcName, voi
     _functions[funcID] = func;
 }
 
-void opencog::setLocalFunc(std::string funcName, Handle* (*func)(AtomSpace *, Handle*))
+void opencog::setLocalSchema(std::string funcName, Handle* (*func)(AtomSpace *, Handle*))
 {
     LibraryManager::setLocalFunc("", funcName, reinterpret_cast<void*>(func));
```
Comment (ngeiswei) [inline opencog/atoms/execution/ExecutionOutputLink.cc]: It looks like the indentation type of this file is tab, while you're using space.

File: tests/atoms/GroundedSchemaLocalUTest.cxxtest [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+using namespace opencog;
+
+#define N as->add_node
+#define L as->add_link
+
+class GroundedSchemaLocalUTest: public CxxTest::TestSuite
+{
+private:
+    AtomSpace* as;
```
Comment (ngeiswei) [inline tests/atoms/GroundedSchemaLocalUTest.cxxtest]: It's really just my personally taste here and it shouldn't influence other's dev choice but I think tab indentation (or even better tab for indentation and space of alignment) is really cool! And it seems new files tend to be tab indented (well probably cause both Linas and I prefer tabs over space)... But again, people's choices...

File: tests/atoms/GroundedSchemaLocalUTest.cxxtest [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+
+    Handle eol =
+        L(EXECUTION_OUTPUT_LINK,
+            N(GROUNDED_SCHEMA_NODE, "lib:\\safe_car"),
+            L(LIST_LINK,
+                N(CONCEPT_NODE, "Arg1"),
+				N(CONCEPT_NODE, "Arg2"),
+				N(CONCEPT_NODE, "Arg3"))
+			);
+	Handle result = ((ExecutionOutputLink *)(eol.atom_ptr()))->execute(as, false);
```
Comment (ngeiswei) [inline tests/atoms/GroundedSchemaLocalUTest.cxxtest]: Though this one is badly indented (or rather, I would argue, the rest is!).

**Comment by Necr0x0Der** [comment]
Hmm... there indeed a mixture of tabs and spaces. It already occurs in the existing files, but here it somehow happens to be within the same piece of code. It looked nice in my editor, so I didn't notice this...
I can fix the indentation in the next commit with GroundedPredicate "lib:" API.

**Comment by ngeiswei** [comment]
Yeah, you should set a tab width different than 4 spaces (I have set it to 3, that's enough to stand out).

**Comment by ngeiswei** [comment]
@Necr0x0Der It's better to fix it in that PR, it's not a good practice to introduce bad stuff (even as minor as bad indentation) with the intension of fixing them later, unless you really have no other choice.

If you push more commits on that branch, github will take them into account, even if they overwrite existing commits, BTW.

**Comment by Necr0x0Der** [comment]
And to have a dead and buggy code is a good practice, yeah... I wonder why people with write access use different rules for themselves... I'm going to start working on GroundedPredicate Friday morning, or not start working on it at all, and switch to another task while intendant fix will wait for CI and approval...

**Comment by ngeiswei** [comment]
Relax, man, I'm just trying to help and avoid adding more mess than there already is.

To your credit, upon further inspection, this file is a mishmash of tabs and spaces so it was indeed hard to follow the "correct" indentation style (actually, by looking at the git history one can guess that the correct style is tab, but you're obviously not expected to go that far).

The rest of the PR looks good, thanks.

**Comment by ngeiswei** [comment]
A piece of advice though, try to stay technical and to the point as possible, that would serve you well everywhere, not just here...

**Comment by ngeiswei** [comment]
Oh, actually the bad indentation is within your own file! So you really have no excuse, jeez... Anyway, it's just bad indent, no big deal, but... You'd be better off striving for your own self-improvements instead of dumping the responsibilities to others (it's a service you are doing to them, not to you! I hope you can see that).

**Comment by Necr0x0Der** [comment]
Did I blame anyone for anything? Did I try to dump the responsibilities to others? I believe, no. I just said that it will be more convenient for me to fix this in the next commit, and I will have no time to wait until next CI pass and PR approval, so in this case I will just fix indentation and will not work on further improvements. And I also noticed that there are many pieces of code in opencog like BROKEN_CODE'ed lines, which neither you nor Linas would approve as PR, so I wondered how they appeared in the repo, and why do you prefer to have this broken code instead of working one. That's all. Yes, it's my fault that I copypasted some code from other repo files and changed it without noticing that it had another type of indentation that I used. I'd be more careful next time. But your 'pieces of advice' regarding self-improvement, etc. are really irrelevant and somewhat insulting.

**Comment by ngeiswei** [comment]
If my piece of advice is irrelevant, it's great, you look like a smart guy, so I guess it likely is... Maybe I felt a bit triggered on that one (hmm, maybe I'm the one who should be striving for self-improvements ;-) ).

Regarding old broken code, OpenCog base have been developed across many years, and was proprietary before it got released by Ben over 10y ago, authors have changed, rules have changed, C++ has evolved, etc, so there's a lot of cruft.

Regarding being blocked by a PR is the waiting, just some technical points
1. As I said, you can still push commits on the PR *after* it has been create (and these commits can even overwrite the previous ones)
2. If what you need isn't merged to the master, you can still have git apply the PR to your local master without touching the history, allowing to keep building upon your work before it's been merged.

Anyway, it's all good, let's move on...

**Comment by ngeiswei** [comment]
@Necr0x0Der you don't need to fix indent, I already did. After reading your comment https://github.com/singnet/atomspace/pull/17#issuecomment-423057825 I wasn't sure what to expect from you so I went ahead, yeah, I might have misread something here, textual communication has that sort of problems, not being able to see the face of the person can lead to that sort of misunderstanding (I tend to stay only technical with that sort of medium to avoid that sort of misunderstanding, but even that doesn't always work perfectly).

**Comment by ngeiswei** [comment]
About merging standard, most of the code you see has been developed on https://github.com/opencog where a lot of people have merging rights (many more than here on https://github.com/singnet), so when something is merged it actually often passes under the radar of Linas and I, and needless to say, we're fallible (and top of that busy), so broken stuff get introduced (including of course by Linas and I) regularly.  We're trying to mitigate that here on singnet by having a smaller group of people with merging writes and continuous integration. We'll see how that works out...

**Comment by Necr0x0Der** [comment]
Yes, that's true that I could create a branch, and work on GroundedPredicate within it while the previous PR is approved, and then merge singnet/atomspace to my fork master, and then merge this branch to master, and it would take less additional time than arguing with you about nothing... but I didn't expect that we would get to this point. Sorry for this. Thanks for fixing the indentation. Let's move on.

**Comment by ngeiswei** [comment]
Sorry as well, it's all good.

**Comment by ngeiswei** [comment]
Just, for the record, this exchange is rather funny on the hindsight. Necr0x0Der turns out to be Alexey, but in my mind he was a newcomer trying to gain some practice. That doesn't mean we don't have lessons to take from it, but it all makes it a lot less serious, and frankly even completely hilarious (I'm still laughing of it). :-D
